### DISCUSSION TITLE
Can't use `refetchInterval` and `setQueryData` together

### DISCUSSION THREAD

**Discussion by amirhhashemi** [TanStack/query#5132] [Q&A] [answered]
I have a query that should be re-fetched every X seconds. The solution is to use `refetchInterval`. But I noticed that calling `setQueryData` resets the `refetchInterval` timer.

In my case, I have a websocket connection that updates part of a query data every 100 milliseconds. If the `refetchInterval` is larger than 100ms, the `queryFn` will never re-run. But I want the `queryFn` to re-run because the websocket only updates part of the query data and I want to keep the rest of it in sync with the server using `refetchInterval`. Our API socks and this is probably an edge case, but I have no choice but dealing with it. Is there a solution?

**Comment by louis-young** [comment, marked as answer]
Hi 👋 

This is an interesting one. As far as I'm aware, the interval is intentionally reset when the query is programmatically updated as this indicates that the query has just been updated. This usually makes sense, as you usually wouldn't want to programmatically update a query and then run the query function 1ms later, for example.

I wonder if you might need to implement this imperatively, or if there's some way to disable this behaviour that I'm not aware of. 

Sorry I can't be of more help but hopefully someone who knows more about the system than me can shed some light 🙂

**Reply by amirhhashemi** [reply]
Hi.
Thanks for the insight.
I think I have to somehow implement it myself, as I couldn't find a way to disable this behavior. I'm waiting for maybe a core team member to confirm it's not possible and hopefully give me some guidance about how to implement it.

**Reply by TkDodo** [reply]
there is no distinction where data comes from: if new data winds up in the cache, it resets the interval. That is on purpose.

**Reply by amirhhashemi** [reply]
> there is no distinction where data comes from: if new data winds up in the cache, it resets the interval. That is on purpose.

Ok. Thank you for the confirmation. 

I ended up saving the query data in a state and update that instead of the query cache.

**Reply by jurosh** [reply]
Is there way to work-around this ? I have data coming from WebSocket which should only update single property of React-Query dataset, so I still want to keep interval to do full re-fetching.

**Comment by lucian55** [comment]
I also need this function, is there any way to do it?
